edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentOperator.cs

```diff
@@ -51,36 +51,23 @@ public void Stop()
 
         public void Dispose() => this.Stop();
 
-        void StartListEdgeDeployments() =>
+        void StartListEdgeDeployments(CancellationTokenSource shutdownCts) =>
             this.client.ListNamespacedCustomObjectWithHttpMessagesAsync(KubernetesConstants.EdgeDeployment.Group, KubernetesConstants.EdgeDeployment.Version, this.deviceNamespace, KubernetesConstants.EdgeDeployment.Plural, watch: true)
-                .ContinueWith(this.OnListEdgeDeploymentsCompleted);
+                .ContinueWith(this.OnListEdgeDeploymentsCompleted, shutdownCts);
 
-        async Task OnListEdgeDeploymentsCompleted(Task<HttpOperationResponse<object>> task)
+        async Task OnListEdgeDeploymentsCompleted(Task<HttpOperationResponse<object>> task, object shutdownCtsObject)
         {
             HttpOperationResponse<object> response = await task;
+            Option<CancellationTokenSource> shutdownCts = Option.Maybe(shutdownCtsObject as CancellationTokenSource);
 
             this.operatorWatch = Option.Some(
                 response.Watch<EdgeDeploymentDefinition, object>(
-                    onEvent: async (type, item) => await this.EdgeDeploymentOnEventHandlerAsync(type, item),
-                    onClosed: () =>
-                    {
-                        Events.EdgeDeploymentWatchClosed();
-
-                        // get rid of the current edge deployment watch object since we got closed
-                        this.operatorWatch.ForEach(watch => watch.Dispose());
-                        this.operatorWatch = Option.None<Watcher<EdgeDeploymentDefinition>>();
-
-                        // kick off a new watch
-                        this.StartListEdgeDeployments();
-                    },
-                    onError: (ex) =>
-                    {
-                        Events.EdgeDeploymentWatchFailed(ex);
-                        throw ex;
-                    }));
+                    onEvent: async (type, item) => await this.EdgeDeploymentOnEventHandlerAsync(type, item, shutdownCts),
+                    onClosed: () => this.RestartWatch(shutdownCts),
+                    onError: (ex) => this.HandleError(ex, shutdownCts)));
         }
 
-        internal async Task EdgeDeploymentOnEventHandlerAsync(WatchEventType type, EdgeDeploymentDefinition item)
+        internal async Task EdgeDeploymentOnEventHandlerAsync(WatchEventType type, EdgeDeploymentDefinition item, Option<CancellationTokenSource> shutdownCts)
```

Comment:
Why use Option<CancellationTokenSource> and not only CancellationTokenSource? In some places, you are using CancellationTokenSource and in others, you are using Option<CancellationTokenSource>. Any specific reason to do this?

Reply:
I'm using the Option type to manage the potentially null object passed into `OnListEdgeDeploymentsCompleted`.  Logically, it shouldn't ever be null, but I think this is once again me being overly cautious.